Caching readers in storage layer

mmaslankaprv opened the pull request:
## Cover letter

Disk log readers are underlie by `seastar::file_input_stream` which dispatches background read-aheads to speed up reading. In previous implementation we throw away the reader even tho the underlying input stream could be reused for next read. 

Introduced readers cache in storage layer. The cache allow us to reuse readers and prevents I/O operation wasting. 

Fixes: #955

## Release notes

- Major read performance improvement for non cached reads.

mmaslankaprv commented:
I see that some of the tests in debug have failed, going to fix it soon.

senior7515 commented on src/v/storage/segment.cc:
```diff
@@ -202,15 +203,35 @@ ss::future<> segment::release_appender() {
+                             * we have to use 'spin lock' approach in here singe
+                             * we do not want to make
+                             */
+                            if (_destructive_ops.try_write_lock()) {
```
should the `try_write_lock()` return the units too  so you can do `if( auto units = try_lock() )

mmaslankaprv replied:
There is a comment on this issue, just few lines above. It is `seastar::semaphore` api. There is no `try_get_units`.

Lazin commented on src/v/storage/segment.cc:
```diff
@@ -202,15 +203,35 @@ ss::future<> segment::release_appender() {
+                            }
+                            // Notice that we do not need abortable sleep in
+                            // here since wait time is really short
+                            return ss::sleep(std::chrono::milliseconds(5))
```
Maybe it's possible to do `write_lock` in the loop with 5ms timeout instead of `try_write_lock`? The code won't need the explicit sleep call since we will have it in `write_lock` call.

senior7515 replied:
this should also be `sleep_abortable()`

mmaslankaprv replied:
I've added there a comment about this being abortable, it is only 5ms. Wrt `write_lock` it would be perfect but, with `write_lock` inserts a waiter to semaphore wait list, this waiter is expected to consume all units if there is a reader already and no all units are available all further read locks would block.

senior7515 commented on src/v/storage/parser.cc:
```diff
@@ -93,15 +93,19 @@ static ss::future<result<iobuf>> verify_read_iobuf(
-    return read_iobuf_exactly(_input, model::packed_record_batch_header_size)
+    auto f = ss::make_ready_future<result<model::record_batch_header>>(_header);
+    if (_header.header_crc == 0 && _header.size_bytes == 0) {
+      f = read_iobuf_exactly(_input, model::packed_record_batch_header_size)
```
this logic here is hard to read. should we refactor this with coros and then add this logic. 

at the very least we should add a comment.

mmaslankaprv replied:
with coros it looks much better.

senior7515 commented on src/v/storage/parser.cc:
```diff
@@ -185,7 +190,10 @@ ss::future<result<stop_parser>> continuous_batch_parser::consume_one() {
-        return consume_records();
+        return consume_records().finally([this] {
+            // reset header
+            add_bytes_and_reset();
```
from a design perspective. i wonder if use the 'peek()' for header bytes? from the underlying stream? i guess we can't peek the stream?

mmaslankaprv replied:
exactly, there is no `peek`

senior7515 commented on src/v/storage/parser.cc:
```diff
@@ -93,15 +93,19 @@ static ss::future<result<iobuf>> verify_read_iobuf(
               return parser_errc::end_of_stream;
           }
-          if (b.size_bytes() != model::packed_record_batch_header_size) {
+          if (
```
these next couple of lines look like a bad formatting? what changed.

mmaslankaprv replied:
I've split changes from formatting, formatting is applied in next commit.

senior7515 commented on src/v/storage/parser.cc:
```diff
@@ -95,90 +95,90 @@ static ss::future<result<iobuf>> verify_read_iobuf(
-            result<model::record_batch_header>>(
-            header_from_iobuf(std::move(b.value())));
-      });
+        f = read_iobuf_exactly(_input, model::packed_record_batch_header_size)
```
i think we should try a coro refactor and see how it does. some of these levels became pretty nested

senior7515 commented on src/v/storage/log_reader.h:
```diff
@@ -71,6 +71,8 @@ class skipping_consumer final : public batch_consumer {
     void print(std::ostream&) const override;
 
 private:
+    void update_expected_next_batch(const model::record_batch_header&);
```
hmmm this is a tricky change. seems ok, but hard to test.

senior7515 commented on src/v/storage/log_reader.cc:
```diff
@@ -268,6 +269,17 @@ log_reader::do_load_slice(model::timeout_clock::time_point timeout) {
+      _config.start_offset > _config.max_offset
+      || _config.bytes_consumed > _config.max_bytes || _config.over_budget) {
+        set_end_of_stream();
+        return ss::make_ready_future<storage_t>();
```
the destructor should have a vassert() that it was closed

mmaslankaprv replied:
it is already there in `log_reader` destructor

senior7515 commented on src/v/storage/log_reader.cc:
```diff
@@ -268,6 +269,17 @@ log_reader::do_load_slice(model::timeout_clock::time_point timeout) {
         return _iterator.close().then(
           [] { return ss::make_ready_future<storage_t>(); });
     }
+    /**
```
something this doesn't take into account is that the file input stream does prefetching. that prefetching consumes memory. I feel like we need to have a bounded set of readers or a shared pool of them or smth.

senior7515 commented on src/v/storage/probe.cc:
```diff
@@ -114,4 +115,44 @@ void probe::add_initial_segment(const segment& s) {
     _partition_bytes -= s.reader().file_size();
 }
+
+void readers_cache_probe::setup_metrics(const model::ntp& ntp) {
```
I think we should call `segment_iterator_cache` now that readers are stateful, they are more like iterators than readers (simle) ? Might we worth some renaming here.

senior7515 commented on src/v/storage/readers_cache_probe.h:
```diff
@@ -0,0 +1,28 @@
+#pragma once
```
should we introduce the probe in a separate commit ?

senior7515 commented on src/v/storage/readers_cache.cc:
```diff
@@ -0,0 +1,195 @@
+std::optional<model::record_batch_reader>
+readers_cache::get_reader(const log_reader_config& cfg) {
+    vlog(stlog.trace, "{} - trying to get reader for: {}", _ntp, cfg);
+    auto it = std::find_if(_readers.begin(), _readers.end(), [cfg](entry& e) {
```
can you add a comment here about why this is a linear search? 

should the datastructure be a queue? every time you add it you push to the back but often pop from the front ?

mmaslankaprv replied:
It is an intrusive list, I will add a comment about linear search

senior7515 commented on src/v/storage/readers_cache.cc:
```diff
@@ -0,0 +1,195 @@
+    it->locked = true;
+    it->reader->reset_config(cfg);
+    _probe.cache_hit();
+    return it->make_cached_reader(this);
```
comment on this line?

senior7515 commented on src/v/storage/readers_cache.h:
```diff
@@ -0,0 +1,152 @@
+#include <seastar/core/coroutine.hh>
+
+namespace storage {
+/**
```
this makes sense, but should we have a little RFC doc w/ a design pic here ?

senior7515 commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -855,33 +896,35 @@ ss::future<> disk_log_impl::do_truncate_prefix(truncate_prefix_config cfg) {
-           * Then delete all segments (potentially including the active segment)
-           * whose max offset falls below the new starting offset.
+           * Then delete all segments (potentially including the active
+           * segment) whose max offset falls below the new starting offset.
```
this looks like a formatting error ?

senior7515 commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -893,9 +936,10 @@ ss::future<> disk_log_impl::do_truncate_prefix(truncate_prefix_config cfg) {
-           * between the commited and dirty offsets. When this occurs we pay a
-           * small penalty to resolve the inconsistency by flushing the log and
-           * advancing the committed offset to be the same as the dirty offset.
+           * between the commited and dirty offsets. When this occurs we pay
```
i don't get what happened here w/ all the comments.

senior7515 commented on src/v/config/configuration.cc:
```diff
@@ -41,6 +42,12 @@ configuration::configuration()
       "256MiB)",
       required::no,
       256_MiB)
+  , readers_cache_eviction_timeout_ms(
```
this is out of order, you need to reorder this commit.

senior7515 reviewed (commented):
I don't feel i understand it fully to approve this PR. I think it needs a detailed walk through with @dotnwat and @rystsov to reason about it. 

I took a first pass at the review.

mmaslankaprv commented:
After yesterday chat with @dotnwat I've introduced following changes comparing to v1:
- removed eviction loop triggered by semaphore from readers cache, now disposing readers is done in background
- Using separate `intrusive_list`s for readers in use and available one instead of `bool locked` flag.
- Split `accept_batch_start` from `consume_batch_start` in `batch_consumer` 
- Removed tail recursion from `consume_header` in parser

Additionally:
- replaced `consume_result` variant with simple enum.

dotnwat commented on src/v/storage/readers_cache.cc:
```diff
@@ -0,0 +1,252 @@
+    _readers.clear_and_dispose([](entry* e) {
+        delete e; // NOLINT
+    });
+    co_return;
```
fyi co_return not needed at end of coroutines returning ss::future<>

dotnwat commented on src/v/storage/readers_cache.h:
```diff
@@ -0,0 +1,140 @@
+          : _e(e)
+          , _cache(c) {
+            _e->_hook.unlink();
+            _cache->_in_use.push_back(*_e);
```
seems like this should happen at the call site that creates the entry guard.

dotnwat commented on src/v/storage/readers_cache.h:
```diff
@@ -0,0 +1,140 @@
+            _e->_hook.unlink();
+            _cache->_in_use.push_back(*_e);
+        }
+        ~entry_guard() {
```
noexcept?

dotnwat commented on src/v/storage/readers_cache.cc:
```diff
@@ -0,0 +1,252 @@
+}
+
+void readers_cache::dispose_in_background(entry* e) {
+    (void)ss::with_gate(_gate, [this, e] {
```
since an entry is in a reader wrapper owned by a independent fiber, can the reader cache shutdown asynchronously to the reader fiber destroying its reader and invoking this via the entry destructor, resulting in with_gate throwing?

mmaslankaprv replied:
yes, I've added correct handling for `gate_closed_exception`

dotnwat commented on src/v/storage/readers_cache.cc:
```diff
@@ -0,0 +1,252 @@
+    for (auto& r : _in_use) {
+        co_await r.reader->finally();
+    }
+    _readers.clear_and_dispose([](entry* e) {
```
why not also for `_in_use` list?

mmaslankaprv replied:
after all readers are closed `_in_use` is empty

dotnwat commented on src/v/storage/segment.cc:
```diff
@@ -202,15 +203,35 @@ ss::future<> segment::release_appender() {
                      a = std::move(a),
                      c = std::move(c),
                      i = std::move(i)]() mutable {
-                        return write_lock().then(
```
I really don't understand the deadlock scenario described in the commit message. could you provide some more detail about the scenario? (hint: it's hard to know what `we` is, for instance, in `...when we want to create reader...`). is this related to the reader cache or a totally separate bug? I have a hunch that this particular problem can be solved without the repeat loop, but I don't fully understand the scenario yet.

mmaslankaprv replied:
I've added this as a commend and commit message: 
Segment close operation is destructive this makes it to require segment write 
lock. When any of the readers is open for given segment it can not be closed 
without waiting since each reader holds a read lock for the segment. In order to
be able to roll the segment without waiting for all readers using this segment 
to be closed when immediate write lock is impossible we dispatch segment close
operation in background. Background close operation acquires a write lock and 
then closes the segment. 

Seastar `read_write_lock` is backed by the semaphore. Acquiring read lock 
requires at least one unit to be available in the semaphore while acquiring 
write lock requires all of the units to be available. 
(semaphore is initialized with max integer count of units). All of the wait 
requests are serialized into semaphore `_waiters` list. This cause the situation
in which any read lock can not be acquired if write lock request is in the 
`_waiters` list.

Example
```
┌──────┬───────┬──────┬──────┬──────┐
│ read │ write │ read │ read │ read │
└──────┴───────┴──────┴──────┴──────┘
   0       1      2      3      4
```

In this situation after read lock from position 0 is acquired any further read 
locks can not be acquired since write lock request is 'blocking' `_waiters` list.

Both of described rules together may lead to deadlock situation in following 
situation:

1. create reader for segment s1.
2. append entries to log to cause segment roll 
3. create reader for segments s1 & s2. 

In this situation step `3.` will never succeed until reader from step 1. is alive.

Using spin lock approach allow us to wait for the write lock without inserting 
wait request to underlying semaphore `_waiters` list.

dotnwat replied:
Thanks for this comment, it is really amazing!

The flip side of this seems to be that reader can deny the ability for storage to roll logs as they grow, right? How problematic might that be if there are long lived active readers and producers? If that's problematic then it seems that we need a way to guarantee progress on rolling.

Given all the work in this patch series that allows log reader instances to be suspended / reconfigured, would it make sense to have a policy by which readers released their locks when they went back into the reader cache? I think there are some other tricks I think we could do that doesn't have the starvation property.

dotnwat commented on src/v/storage/segment.cc:
```diff
@@ -202,15 +203,35 @@ ss::future<> segment::release_appender() {
+                            }
+                            // Notice that we do not need abortable sleep in
+                            // here since wait time is really short
+                            return ss::sleep(std::chrono::milliseconds(5))
```
I think you can use `ss::later()` instead of setting a timer--but take a look at its implementation I don't exactly know what it does. what you seem to want here is a method to yield, and there are also other reactor interface that seem to allow a preempt to be requested.

mmaslankaprv replied:
`ss::later()` is actually the way to break tail recursion. It indeed yield but it creates a task in a task queue, in this situation it would make unnecessary pressure with empty tasks in a queue.

dotnwat replied:
Doesn't sleeping add a task into the task queue (the future that resolves when the sleep is done?)

dotnwat commented on src/v/storage/readers_cache.cc:
```diff
@@ -0,0 +1,252 @@
+ss::future<> readers_cache::stop() {
+    co_await _gate.close();
+    for (auto& r : _readers) {
+        co_await r.reader->finally();
```
is it really necessary to wait on all the readers.finally? it seems like since cache doesn't ultimately control the lifetime of the user of the reader then this could become problematic if some reader had a long lifetime on shutdown. would an alternative approach work where in the entry_guard you cleared a bit/pointer so that the entyr_guard destructor became a noop? this might be another case where using circular_buffer<lw_shared_ptr> simplifies things.

mmaslankaprv replied:
We would wait for all those readers in `disk_log_impl::close()` since it requires write lock.

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -81,16 +84,20 @@ ss::future<> disk_log_impl::remove() {
-            kvstore::key_space::storage,
-            internal::start_offset_key(config().ntp()));
+
+    return _readers_cache->stop().then(
```
this strikes me as problematic (e.g. unnecessary delays controlled by random readers) to wait on readers. we didn't before, right? so it seems that if the cache is really a cache and not an _owning_ set of readers then it is in principle unnecessary. I added some more comments about this to the cache::stop method in another patch.

mmaslankaprv replied:
we waited  for all those readers before, segment close needs a write lock.

dotnwat replied:
the point was that it adds unnecessary coupling and code. when the cache is stopped it can arrange for all entry guards to become noops (breaking dependency) and then all of this code where we have to wrap shutdown sequences in the stop continuation of the cache reader would seem to go away.

mmaslankaprv replied:
we will require this `finally()` call for each reader that is not in use, additionally when reader is returned from the cache is wrapped in `cached_reader`. It makes the `finally()` call to be noop

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -798,11 +833,13 @@ ss::future<> disk_log_impl::remove_segment_permanently(
-          vlog(stlog.error, "Cannot close segment: {} - {}", e, s);
-      })
-      .finally([this, s] { _probe.segment_removed(); });
+    return _readers_cache->evict_on_segment_remove(s).then([this, s] {
```
same concern as cache::stop. seems like we could invalidate the readers' entry_guard destructor to avoid waiting on the readers' finally future to resolve?

mmaslankaprv replied:
we have to wait since close requires a write lock.

dotnwat replied:
the point was that it adds unnecessary coupling and code. when the cache is stopped it can arrange for all entry guards to become noops (breaking dependency) and then all of this code where we have to wrap shutdown sequences in the stop continuation of the cache reader would seem to go away.

dotnwat commented on src/v/storage/segment_utils.cc:
```diff
@@ -470,42 +480,42 @@ ss::future<> self_compact_segment(
-                __builtin_unreachable();
-            }
-        })
+      .then([idx_path, s, cfg, &pb, &readers_cache](
```
this reformatting needs to be in a separate patch please

dotnwat reviewed (changes requested):
the overall design looks great and the results of this seem really impressive!

the early patches in the series that make changes to the log parser are really well designed, but a few of those patches that evolve the parser seem to represent intermediate steps with problematic semantics (e.g. physical_offset update). the cache looks good, but I'm wondering if it could be simplified quite a lot by using shared ptr instead of intrusive list.

mmaslankaprv commented on src/v/storage/parser.cc:
```diff
@@ -108,16 +107,13 @@ ss::future<result<stop_parser>> continuous_batch_parser::consume_header() {
+            co_return stop_parser::yes;
+        case batch_consumer::consume_result::accept_batch:
+            co_return stop_parser::no;
+        case batch_consumer::consume_result::skip_batch:
```
exactly

mmaslankaprv commented on src/v/storage/parser.cc:
```diff
@@ -93,88 +94,81 @@ static ss::future<result<iobuf>> verify_read_iobuf(
+            auto should_skip = std::get<skip_batch>(ret);
+            // do not skip batch
+            if (likely(should_skip == skip_batch::no)) {
+                _physical_base_offset += _header.size_bytes;
```
you are right, I will make it unconditional.

mmaslankaprv commented on src/v/storage/parser.cc:
```diff
@@ -93,88 +94,81 @@ static ss::future<result<iobuf>> verify_read_iobuf(
+    /**
+     * we use a loop to prevent using tail recursion
+     **/
+    for (;;) {
```
I am not sure if this is a generator that would benefit from `co_yield` there are multiple exit points in this method.

mmaslankaprv commented on src/v/storage/parser.cc:
```diff
@@ -93,88 +94,81 @@ static ss::future<result<iobuf>> verify_read_iobuf(
+
+    if (unlikely(header.header_crc == 0)) {
+        // happens when we fallocate the file
+        co_return parser_errc::fallocated_file_read_zero_bytes_for_header;
```
you are right, in the original code there were 2 conditions. They both checked the same. The crc equal to 0 really mean that parser entered fallocated area in a file. Both error codes are treated the same way down the stream. I've fixed the behavior of original code.

mmaslankaprv commented on src/v/storage/parser.cc:
```diff
@@ -98,11 +98,13 @@ ss::future<result<stop_parser>> continuous_batch_parser::consume_header() {
-        auto r = co_await read_header();
-        if (!r) {
-            co_return r.error();
+        if (_header.header_crc == 0 && _header.size_bytes == 0) {
```
changed `_header` to be `optional`.

mmaslankaprv commented on src/v/storage/parser.cc:
```diff
@@ -179,7 +181,7 @@ ss::future<result<stop_parser>> continuous_batch_parser::consume_one() {
             return ss::make_ready_future<result<stop_parser>>(st.value());
         }
-        return consume_records();
+        return consume_records().finally([this] { add_bytes_and_reset(); });
```
Change it to `.then` to keep previous semantincs.

mmaslankaprv commented on src/v/storage/log_reader.cc:
```diff
@@ -38,14 +34,28 @@ batch_consumer::consume_result skipping_consumer::consume_batch_start(
           _expected_next_batch,
           header.base_offset()));
```
`accept_batch` may be called more than once with the same batch. We talked about this during our interactive review, your proposal was to split accepting batch from consuming and handle an assertion this way.

mmaslankaprv replied:
this commit doesn't yet introduce the `accept_batch` method

mmaslankaprv commented on src/v/storage/log_reader.cc:
```diff
@@ -38,14 +34,28 @@ batch_consumer::consume_result skipping_consumer::consume_batch_start(
+    /**
+     * Check if we have to skip the batch
+     */
+    _expected_next_batch = header.last_offset() + model::offset(1);
```
I think this comment shouldn't belong here. I will make `accept` method `const`. Do we need to pass skipped batch to consumer ?

mmaslankaprv replied:
I got it now, it is nice idea, let's do this.

mmaslankaprv commented on src/v/storage/readers_cache.cc:
```diff
@@ -0,0 +1,252 @@
+
+model::record_batch_reader
+readers_cache::put(std::unique_ptr<log_reader> reader) {
+    if (reader->lease_range_base_offset() < model::offset{0}) {
```
this occur when start_offset is greater than max offset.

mmaslankaprv commented on src/v/storage/readers_cache.cc:
```diff
@@ -0,0 +1,252 @@
+    vlog(stlog.trace, "{} - trying to get reader for: {}", _ntp, cfg);
+    intrusive_list<entry, &entry::_hook> to_evict;
+    /**
+     * We use linear search since _readers intrusive list is small.
```
I was considering this, I think that we have already some experience with intrusive lists and additionally `touch` operation doesn't involve any allocation. Maybe we would have to move it to be shard level cache common for all ntps on the same shard. (similar to chunk cache)

senior7515 commented:
> the overall design looks great and the results of this seem really impressive!
> 
> the early patches in the series that make changes to the log parser are really well designed, but a few of those patches that evolve the parser seem to represent intermediate steps with problematic semantics (e.g. physical_offset update). the cache looks good, but I'm wondering if it could be simplified quite a lot by using shared ptr instead of intrusive list.

@mmaslankaprv i didn't see this addressed.

mmaslankaprv commented:
> > the overall design looks great and the results of this seem really impressive!
> > 
> > the early patches in the series that make changes to the log parser are really well designed, but a few of those patches that evolve the parser seem to represent intermediate steps with problematic semantics (e.g. physical_offset update). the cache looks good, but I'm wondering if it could be simplified quite a lot by using shared ptr instead of intrusive list.
> 
> @mmaslankaprv i didn't see this addressed. 

We talked about this with Noah during interactive review.

dotnwat commented on src/v/storage/kvstore.cc:
```diff
@@ -536,20 +537,27 @@ void kvstore::replay_segments_in_thread(segment_set segs) {
       header.base_offset,
       _store->_next_offset);
     _header = header;
-    return batch_consumer::consume_result::accept_batch;
```
@mmaslankaprv  needs to go in previous commit?

dotnwat replied:
oh never mind. the enum has the same name!

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -723,7 +753,15 @@ disk_log_impl::make_reader(log_reader_config config) {
+        return ss::make_ready_future<model::record_batch_reader>(
+          std::move(empty));
+    }
+    return make_cached_reader(config);
```
do we need to have cached readers that are built based on time queries (like below this comment?)

dotnwat commented on src/v/storage/segment_utils.cc:
```diff
@@ -407,16 +411,20 @@ ss::future<> do_self_compact_segment(
+      .then([s, &readers_cache](storage::index_state idx) {
+          return readers_cache.evict_segment_readers(s).then(
+            [s,
+             idx = std::move(idx)](readers_cache::range_lock_holder) mutable {
```
why is holder's lifetime not prolonged?

mmaslankaprv replied:
we can relay on the fact that `write_lock` is synchronous up to the point where waiter is put to the `_waiters` list

dotnwat commented on src/v/storage/segment_utils.cc:
```diff
@@ -470,42 +481,42 @@ ss::future<> self_compact_segment(
+                });
+          }
+          default:
+              __builtin_unreachable();
```
nit: builtin unreachable should not be in the default case. it prevents compiler from catching missing enums.

mmaslankaprv replied:
this change comes from reformatting, I will move this into different commit

dotnwat reviewed (commented):
nice work on the new locking stuff and getting rid of the loop and sleep. i still have reservations about what's happening when we stop the reader cache. it looks problematic, but maybe I'm missing something. other than that and a couple questions I think this is good to go.

dotnwat commented on src/v/storage/readers_cache.cc:
```diff
@@ -0,0 +1,295 @@
+    return e.make_cached_reader(this);
+}
+
+ss::future<> readers_cache::stop() {
```
I see now that many users in disk log reader save the holder in a finally continuation, but that doesn't seem to obviously synchronize with stopping the cache. the lifetime rules seem really complicated. is that explained some where?

mmaslankaprv replied:
Readers cache is closed before when `disk_log_impl` is closed. This way we prevent any operation to hold the cache reference before it is destroyed. An instance of `readers_cache` is owned by `disk_log_impl`

mmaslankaprv commented on src/v/storage/log_reader.h:
```diff
@@ -153,6 +153,32 @@ class log_reader final : public model::record_batch_reader::impl {
+     * read all locked segments already
+     */
+    bool is_reusable() const {
+        return _iterator.current_reader_seg != _lease->range.end()
```
good point, an iterator check is redundant in here.

mmaslankaprv commented on src/v/storage/readers_cache.cc:
```diff
@@ -0,0 +1,295 @@
+
+model::record_batch_reader
+readers_cache::put(std::unique_ptr<log_reader> reader) {
+    if (reader->lease_range_base_offset() < model::offset{0}) {
```
no, readers are working in storage offsets domain

mmaslankaprv commented on src/v/storage/readers_cache.cc:
```diff
@@ -0,0 +1,295 @@
+    // range locked, do not insert into the cache
+    if (lock_it != _locked_offset_ranges.end()) {
+        return model::record_batch_reader(std::move(reader));
+    }
```
It would work, I didn't want to introduce yet another entity kept in the reader, I thought that with ranges we reduce coupling.

dotnwat replied:
yeh I think its ok. IMO offset ranges are harder to reason about and have edge cases. but I think they are equivalent.

mmaslankaprv commented on src/v/storage/readers_cache.cc:
```diff
@@ -0,0 +1,295 @@
+ss::future<readers_cache::range_lock_holder>
+readers_cache::evict_prefix_truncate(model::offset o) {
+    vlog(stlog.debug, "{} - evicting reader prefix truncate {}", _ntp, o);
+    offset_range range{model::offset::min(), o};
```
no, it is entirely in storage offset domain

mmaslankaprv commented on src/v/storage/readers_cache.cc:
```diff
@@ -0,0 +1,295 @@
+        co_await r.reader->finally();
+    }
+    for (auto& r : _in_use) {
+        co_await r.reader->finally();
```
I think your points are valid. I've changed the `readers_cache::stop()` implementation to wait for all in_use readers.
